config_v2.md

```diff
@@ -0,0 +1,225 @@
+
+## Non-goals of this document:
+
+* Describing or even caring about the migration process. It can be done at a later stage.
+
+* Caring about looking like our old API about configs. Legacy API design should not prevent us from pushing a better API.
+
+* Handling seeding. The seeding has been separated from the configuration process with the PR [Making the seeding more intuitive](https://github.com/IDSIA/sacred/pull/615)
+
+
+## API description:
+
+The goal is to make the API easier to understand.
+
+* There is no more `Ingredient`. 
+
+* There is no config scopes.
+
+* Ingredients are replaced by functions which can be activated either from the command line or when creating the Experiment.
+
+* No local variable gathering, injections or override. In short, no black magic.
+
+* Classes and functions can be put in the config, but only the full name will be saved with the Observers.
+
+
+### Basic example:
+
+
+```python
+import sacred
+
+
+configuration = dict(batch_size=32, dataset_size=10_000, nb_epochs=50)
+
+
+ex = sacred.Experiment('my_pretty_experiment',
+                       config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**ex.config)
+```
+
+
+### Example with the log and run object:
+
+```python
+import sacred
+
+
+configuration = dict(batch_size=32, dataset_size=10_000, nb_epochs=50)
+
+
+ex = sacred.Experiment('my_pretty_experiment',
+                       config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs, run, logger):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**ex.config, run=ex.current_run, logger=ex.logger)
+```
+
+
+### Example with delayed evaluation of a config variable:
+
+
+This handles the case where a configuration value, if not set should be evaluated based on other values present in the config.
+
+If the `Delayed` object is present even after all overrides from the command lines and other, it is evaluated before starting the Experiment.
+
+```python
+import sacred
+from sacred import Delayed
+
+
+configuration = dict(
+    batch_size=32, 
+    dataset_size=Delayed(lambda config: config['batch_size'] * 100), 
+    nb_epochs=50
+)
+
+
+ex = sacred.Experiment('my_pretty_experiment',
+                       config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**ex.config)
+```
+
+For example:
+
+```bash
+python my_main.py with batch_size=60  # dataset_size is 6000 here
+python my_main.py with dataset_size=5000  # dataset_size is 5000 here
+```
+
+
+##### Api of the Delayed object:
+
+```python
+class Delayed:
+    def __init__(self, evaluation_function, priority=0):
```

Comment:
`priority` seems like an important feature for some use cases but I do worry it could result in very hard to understand configs.

I think the typical use case would be when one config depends on the value of a previous config. Maybe configs could explicitly declare what config value(s) or previous config(s) they depend on. This would produce a DAG, and then Sacred could evaluate the configs in topological order.

Reply:
True. It seems this `priority` will be useful on only very specific and rare cases. @Higgcz also mentioned another workaround. Let's drop it. If we see that there is a demand for it in the future, we'll do something.